projects/stretch_continual/stretch_continual/envs/stretch_demo_base_env.py

```diff
@@ -0,0 +1,312 @@
+                    if try_id == max_tries - 1:
+                        print(
+                            f"Failed to open or read {output_path} with error {e} after {max_tries} attempts."
+                        )
+                        raise e
+
+                    print(
+                        f"Failing to open or read {output_path} with error {e}. Retrying..."
+                    )
+
+        if cache:
+            demo_data = CacheContainer(self._trajectory_cache)
+        else:
+            demo_data = h5py.File(output_path, "r")
+
+        return demo_data
+
+    def randomly_select_traj_from_dir(self, directory, only_key_frames, cache):
+        max_tries = 25
+
+        for try_id in range(max_tries):
+            try:
+                with tempfile.NamedTemporaryFile(
+                    dir=directory
+                ) as temp_aggregation_file:
+                    with self.load_all_h5_from_dir(
+                        directory,
+                        only_key_frames,
+                        temp_aggregation_file,
+                        cache=cache,
+                    ) as h5_file:
+                        aggregated_h5 = h5_file["linked_files"]
+                        traj_counts = [
+                            (file_key, len(item.keys()))
+                            if item is not None
+                            else (None, 0)
+                            for file_key, item in aggregated_h5.items()
+                        ]
+                        total_trajs = np.array(
+                            [count[1] for count in traj_counts]
+                        ).sum()
+                        traj_id = np.random.randint(total_trajs)
+                        curr_id = 0
+                        traj = None
+
+                        for file_id, count in traj_counts:
+                            # If this file contains our selected trajectory (i.e. it's within the count for this file), grab it
+                            if curr_id + count > traj_id:
+                                traj = aggregated_h5[file_id][f"{traj_id - curr_id}"]
+                                break
+
+                            curr_id += count
+
+                break
+            except (BlockingIOError, OSError, KeyError) as e:
+                time.sleep(0.1)
+
+                if try_id == max_tries - 1:
+                    print(f"Ran out of re-attempts to load the h5.")
+                    raise e
+
+        return traj
+
+    def gripper_fk(self, model, pose):
+        ee_pos, ee_rot = model.manip_fk(pose)
+        return ee_pos, ee_rot
+
+    def gripper_ik(self, model, pos, rot, current_joints):
+        ros_pose, ik_success, ik_debug_info = model.manip_ik(
+            (pos, rot), q0=current_joints
+        )
+
+        # Prevent full rotations of the wrist...
+        raw_roll = ros_pose[HelloStretchIdx.WRIST_ROLL]
+        positive_roll = raw_roll % (2 * np.pi)
```

Comment:
what is going on here?

Reply:
If I recall correctly, sometimes around the 0 point (i.e. the 2pi point) instead of, say, 0.05 I was observing that it would predict 2pi+0.05, which would result in an extra roll. If there is something in its grippers (e.g. the handle of an oven) this was...not great. It won't prevent "legitimately" crossing the boundary, but reduces the chance of spurious rotations.

(This is what I was trying to convey with "prevent full rotations of the wrist". Let me know if you have suggestions for more clarity. I also am not 100% sure this is still a problem, with the changes/improvements to ik, but I can't easily check right now.)

Reply:
(Oh, I think the "positive roll" comes from the fact that sometimes it would be -0.05 instead of 2pi-0.05, actually. So this is just making it more consistent. 

Btw in python (-2 % 5 == 3) This is not consistent across all languages and is somewhat non-obvious. If this is the source of confusion I can add a comment.)

Reply:
No its ok